pkg/kv/dynamodb/store.go

```diff
@@ -0,0 +1,320 @@
+}
+
+func (s *Store) SetIf(ctx context.Context, key, value []byte, valuePredicate kv.Predicate) error {
+	return s.setWithOptionalPredicate(ctx, key, value, valuePredicate, true)
+}
+
+func (s *Store) setWithOptionalPredicate(ctx context.Context, key, value []byte, valuePredicate kv.Predicate, usePredicate bool) error {
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+
+	item := DynKVItem{
+		PartKey:   s.getPartitionKey(key),
+		ItemKey:   string(key),
+		ItemValue: string(value),
+	}
+
+	marshaledItem, err := dynamodbattribute.MarshalMap(item)
+	if err != nil {
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+
+	input := &dynamodb.PutItemInput{
+		Item:      marshaledItem,
+		TableName: &s.params.TableName,
+	}
+	if usePredicate {
+		if valuePredicate != nil {
+			input.ConditionExpression = aws.String(ItemValue + " = :predicate")
+			input.ExpressionAttributeValues = map[string]*dynamodb.AttributeValue{
+				":predicate": {S: aws.String(string(valuePredicate.([]byte)))},
+			}
+		} else {
+			input.ConditionExpression = aws.String("attribute_not_exists(" + ItemValue + ")")
+		}
+	}
+
+	_, err = s.svc.PutItemWithContext(ctx, input)
+
+	if err != nil {
+		if _, ok := err.(*dynamodb.ConditionalCheckFailedException); ok && usePredicate {
+			return kv.ErrPredicateFailed
+		}
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+	return nil
+}
+
+func (s *Store) Delete(ctx context.Context, key []byte) error {
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+
+	_, err := s.svc.DeleteItemWithContext(ctx, &dynamodb.DeleteItemInput{
+		TableName: aws.String(s.params.TableName),
+		Key:       s.bytesKeyToDynamoKey(key),
+	})
+
+	if err != nil {
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+	return nil
+}
+
+func (s *Store) Scan(ctx context.Context, start []byte) (kv.EntriesIterator, error) {
+	return s.scanInternal(ctx, start, nil)
+}
+
+func (s *Store) scanInternal(ctx context.Context, scanKey []byte, exclusiveStartKey map[string]*dynamodb.AttributeValue) (*EntriesIterator, error) {
+	keyConditionExpression := PartKey + " = :partkey"
+	expressionAttributeValues := map[string]*dynamodb.AttributeValue{
+		":partkey": {
+			S: aws.String(s.getPartitionKey(scanKey)),
+		},
+	}
+	if len(scanKey) > 0 {
+		keyConditionExpression += " AND " + ItemKey + " >= :fromkey"
+		expressionAttributeValues[":fromkey"] = &dynamodb.AttributeValue{
+			S: aws.String(string(scanKey)),
+		}
+	}
+	queryInput := &dynamodb.QueryInput{
+		TableName:                 aws.String(s.params.TableName),
+		KeyConditionExpression:    aws.String(keyConditionExpression),
+		ExpressionAttributeValues: expressionAttributeValues,
+		ConsistentRead:            aws.Bool(true),
+		ScanIndexForward:          aws.Bool(true),
+		ExclusiveStartKey:         exclusiveStartKey,
+	}
+	queryOutput, err := s.svc.QueryWithContext(ctx, queryInput)
```

Comment:
The number of items is indeterminate and is dictated by the items size. `DynamoDB` returns up to ~1MB of data, regardless of the number of items.
In addition the number of items can be limited, using a specified value (which we currently not using).
AFAICT, our tests **does not cover** pagination, but I tested it manually

Reply:
Our tests can cover pagination by design since it's a database internal implementation feature. If they do not cover it, too much code is kept uncovered. We should either add\change the store tests to fit it, or write something designated to DynamoDB.

Reply:
I tried to resize our scan test to breach the pagination of ~1MB and it seems to take ages.
In addition, I don't thing this approach leads to a scalable test, as today it is 1MB but who knows what bears the future
I thought of leveraging the option to specify a limit on the number of entries per each page and so create a real, sustainable pagination.
It can be added as another param to the dsn. WDYT?

Reply:
Done.
Added paging size parameter and used it to force paging in tests.
Also found and fixed a bug in the delete while scan with paging, where deleting the entire last page, while scanning the former page, will lead to a crash, so +1 for that comment!